test/Test/Database/LSMTree/Internal/Chunk.hs

```diff
@@ -0,0 +1,180 @@
+               (map show (buildupPhasesLengths commonChunks))
+               (consumer commonChunks remnant)
+    where
+
+    commonChunks :: [Maybe Chunk]
+    remnant :: Maybe Chunk
+    (commonChunks, remnant) = balingOutput minChunkSize food
+
+    buildupPhasesLengths :: [Maybe Chunk] -> [Int]
+    buildupPhasesLengths []     = []
+    buildupPhasesLengths chunks = List.length buildupOutput :
+                                  buildupPhasesLengths (List.drop 1 followUp)
+        where
+
+        buildupOutput, followUp :: [Maybe Chunk]
+        (buildupOutput, followUp) = span isNothing chunks
+
+-- * Properties to test
+
+prop_contentIsPreserved :: MinChunkSize -> [[Vector Word8]] -> Property
+prop_contentIsPreserved (MinChunkSize minChunkSize) food
+    = withBalingOutput minChunkSize food $ \ commonChunks remnant ->
+      let
+
+          input :: Vector Word8
+          input = concat (List.concat food)
+
+          output :: Vector Word8
+          output = concat (fromChunk <$> catMaybes (commonChunks ++ [remnant]))
+
+      in input == output
+
+prop_noRemnantAfterOutput :: MinChunkSize
+                          -> NonEmptyList [Vector Word8]
+                          -> Property
+prop_noRemnantAfterOutput (MinChunkSize minChunkSize) (NonEmpty food)
+    = withBalingOutput minChunkSize food $ \ commonChunks remnant ->
+      isJust (last commonChunks) ==> isNothing remnant
+
+prop_commonChunksAreLarge :: MinChunkSize -> [[Vector Word8]] -> Property
+prop_commonChunksAreLarge (MinChunkSize minChunkSize) food
+    = withBalingOutput minChunkSize food $ \ commonChunks _ ->
+      all (fromChunk >>> length >>> (>= minChunkSize)) (catMaybes commonChunks)
+
+remnantChunkSizeIs :: (Int -> Bool) -> Int -> [[Vector Word8]] -> Property
+remnantChunkSizeIs constraint minChunkSize food
+    = withBalingOutput minChunkSize food $ \ _ remnant ->
+      isJust remnant ==> constraint (length (fromChunk (fromJust remnant)))
+
+prop_remnantChunkIsNonEmpty :: MinChunkSize -> [[Vector Word8]] -> Property
+prop_remnantChunkIsNonEmpty (MinChunkSize minChunkSize)
+    = remnantChunkSizeIs (> 0) minChunkSize
+
+prop_remnantChunkIsSmall :: MinChunkSize -> [[Vector Word8]] -> Property
+prop_remnantChunkIsSmall (MinChunkSize minChunkSize)
+    = remnantChunkSizeIs (< minChunkSize) minChunkSize
+
+-- * Tests
+
+tests :: TestTree
+tests = testGroup "Test.Database.LSMTree.Internal.Chunk" $
+        [
+            testProperty "Content is preserved"
+                         prop_contentIsPreserved,
+            testProperty "No remnant after output"
+                         prop_noRemnantAfterOutput,
+            testProperty "Common chunks are large"
+                         prop_commonChunksAreLarge,
+            testProperty "Remnant chunk is non-empty"
+                         prop_remnantChunkIsNonEmpty,
+            testProperty "Remnant chunk is small"
+                         prop_remnantChunkIsSmall
+        ]
```

Comment:
That’s true. However, it would be inconsistent with the style I use within this module (and all other modules by me): ordering things from abstract to concrete and – and this is the relevant thing here – from simple to complex, a style that tends to lead to items on the same level (module, `where` clauses, etc.) referencing only previous items in their implementations.

My preference is to leave the module as it is, but if you would strongly like me to change it then I could also change it. What are your thoughts?

Reply:
Meanwhile, I was able to come up with an approach for reordering and slightly regrouping the module’s contents such that the result meets your goal while still being more or less in line with my goals sketched above (in particular, the contents of the individual sections of the module satisfy the relevant ordering constraint exactly). The corresponding changes are introduced by 1d9bf6b5eebbd2a76ef149e62fa85116a1a5d4b9.

Reply:
That looks good to me!